### ISSUE TITLE
Full VPN Support with WSL2

### ISSUE THREAD

**Issue by martencassel** [microsoft/WSL#10380]
**Is your feature request related to a problem? Please describe.**
Many people work from home with WSL2 and connect to a corporate networking using VPN.
The issues has been that either you choose WSL1 or you have to go to the office.

**Describe the solution you'd like**
Making WSL2 network connectivity seamless work with any VPN solution. 

**Describe alternatives you've considered**
The only stable way i've found is https://github.com/sakai135/wsl-vpnkit and Windows 11 Version 22H2 (OS Build 22621.1992). It works but on startup the system will "freeze" for ca 1-2 minutes. 

**Question**
Is VPN support on any of your roadmap ?

**Comment by ascheel** [comment]
https://github.com/deanmcniven/wsl2-vpn-support

Depending on your VPN, the above may work.  It's just a PowerShell script that adds the appropriate routes, giving WSL2 network access the way it should.

**Comment by ThaDaVos** [comment]
I can add that if I use VPN in Windows (before starting WSL 2 - but sometimes also after starting WSL 2) - it also works from within WSL 2 - I used VPN inside Windows and the OpenVPN client, both successfully

**Comment by liam-baker-sm** [comment]
Usually it works because Windows will create A NAT for your WSL2, and route through the VPN into Windows. The problem I have seen is the IP address range used in WSL2 conflicts with internal network ranges i.e. (172.16.0.0/12)

**Comment by gwd666** [comment]
Ok here is my 2 cents (or experience) with this.
I/we work behind (http) proxies 
On-site (ie w/o VPN) all green. 
From home with VPN to corp network -> issues. 

Interesting details! 
How VPN works for "some time". 
WSL2 internet connectivity will work if there is no `vEthernet (WSL2)`  showing up in `netsh interface show interface`  before I connect via VPN. 
Then I will activate or connect via VPN. 
Internet Connectivity will work and `netsh interface show interface` will now show `vEthernet (WSL2)` as `Enabled` in State `Connected`.     
BUT as soon as the VPN (even if only for the blink of an eye) looses connection and therefore immediately reconnects again, there is no way (at least I have not found one) to make the WSL2 internet connectivity work again (ie proxy sever can no longer be found, etc) EXCEPT rebooting the computer (which will remove ie again show no `vEthernet (WSL2)` interface again.    
Things I tried so far to avoid having to reboot/restart the PC.     
`disable/enable vEthernet (WSL2)` in Powershell elevated terminal 
`Restart LxssManager`      
additonal info: this was done in all available combinations of `wsl --shutdown`  and VPN connect/disconnect you can think of or find on the internet. All to no avail - still the only fix -> reboot until next connection loss and reconnect.    
The VPN connection loss and reconnect breaking the WSL2 internet connectivity since starting WSL2 before being connected via VPN (ie create an existing `vEthernet (WSL2)` interface before there's a VPN involved) and then starting VPN connection afterwards will also result in WSL2 internet connectivity being dead.    
   
Looking fwd to solution ideas ...

**Comment by jabulon** [comment]
I've struggled with this for several months. It seems that a recent [update ](https://devblogs.microsoft.com/commandline/windows-subsystem-for-linux-september-2023-update/)to WSL offers experimental features which resolve the issue for me. I'm using WSL2 on Windows 11 with Cisco AnyConnect VPN.
In my environment, I've added the following flags to my C:\Users\_username_\.wslconfig file:
[experimental]
networkingMode=mirrored
dnsTunneling=true

**Comment by joeybinz** [comment]
This solution worked perfectly for me! I have been following this thread for a few months hoping for a solution to this issue and was far too excited that this worked haha. Thanks.

**Comment by liam-baker-sm** [comment]
Thanks for that information, mirrored mode looks interesting.
Our current fix is to set a registry key to move the NAT to `192.168.240.0/20`
This keeps WSL NAT away from our corporate and most home networks.
```
Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Lxss
NatGatewayIpAddress 192.168.240.1
NatNetwork 192.168.240.0/20
```

**Comment by gwd666** [comment]
update
installed wsl2 with "experimental" vpn support
here are the details
```
WSL version: 2.2.1.0
Kernel version: 5.15.150.1-2
WSLg version: 1.0.60
MSRDC version: 1.2.5105
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.25131.1002-220531-1700.rs-onecore-base2-hyp
Windows version: 10.0.19045.4170
``` 
Still same issue - once the VPN gets disconnected (usually due to some network issues I guess - not all network connections are 100% stable a 100% of the time) and re-connects the WSL connections are gone/dead and I need to restart my PC again for them to come alive.

**Comment by dragz** [comment]
I can confirm that networkingMode=mirrored mentioned above seems to solve my issues with VPN and WSL2. I've had two issues issues  with networking/VPN/WSL2 in the six weeks using this setup where ssh and az (Azure cli) stops working while telnet and other network services like DNS works fine. Running the command 

ipconfig /release

seems to resolve the problem.

I'm using Cisco Anyconnect.

**Comment by ygmarchi** [comment]
I've read the dns configuration in my openvpn log and configured /etc/resolv.conf accordingly, after disabling automatic generation in /etc/wsl.conf. That's the only way the worked for me, pretty empirically.

**Comment by madscientist42** [comment]
Five years since these issues showed up.  NO fixes.  (I don't consider needing to reboot it all when you lose the VPN for any reasons as it being fixed...)

I mean, c'mon- VirtualBox manages this one thing right.  And right now even.  It's such that if I thought I could get my employer to sign off on a Linux box with their Win10 Enterprise install living in a VirtualBox sandbox, I'd go that way.

**Comment by seb-slowik** [comment]
I have successfully managed to get the internet working on my WSL2: Ubuntu 22.04.2 LTS a few times, but it was not consistent until a while back. I have identified a series of steps that reliably enable internet access in WSL2 for me. These steps must be followed in the specified order and should work on both Windows 10 and Windows 11, as I discovered this solution over a year ago while using Windows 10:

1. After Windows starts, before connecting to the VPN, launch your WSL2 instance (in my case, Ubuntu 22.04.2 LTS).
2. Run the following command to check for internet access:
`sudo apt update`
3. Connect to the VPN using your client (in my case, "Cisco Secure Client").
4. Open PowerShell (x86) with local administrator privileges and execute the following commands:
`Get-NetAdapter | Where-Object {$_.InterfaceDescription -Match "Cisco AnyConnect"} | Set-NetIPInterface -InterfaceMetric 5500
Restart-Service LxssManager`
5. Open a non-elevated Command Prompt and run:
`wsl --shutdown`
Note: If you did not close the WSL2 window from step 1, it should close after running this command.
6. Restart the WSL2 client, and you should now have internet access while connected to the VPN network remotely.

These steps hopefully help you achieve consistent internet connectivity in WSL2.

**Comment by h471x** [comment]
> I've struggled with this for several months. It seems that a recent [update ](https://devblogs.microsoft.com/commandline/windows-subsystem-for-linux-september-2023-update/)to WSL offers experimental features which resolve the issue for me. I'm using WSL2 on Windows 11 with Cisco AnyConnect VPN. In my environment, I've added the following flags to my `C:\Users\_username_\.wslconfig` file:
> 
> ```
> [experimental]
> networkingMode=mirrored
> dnsTunneling=true
> ```

@jabulon thank you very much, it worked well for me 👌😊

**Comment by harriswes** [comment]
Wondering if anyone has any ideas. I've had luck connecting to internet while on VPN through Ubuntu LTS, but noticed I can't ping any of my floating servers. I looked at the IP and it looks like WSL2 is grabbing my LAN IP instead of my VPN IP when network mirroring. Can anyone think of a simple fix without having to run a bash script everytime I open WSL?

**Comment by dragz** [comment]
On my setup (Win11/Cisco Anyconnect/WSL2/Ubuntu 24.04 LTS) with mirrored networking I have no issues like this. Both the lan ip and the vpn ip is visible from the Ubuntu system. Maybe a routing problem? Do you reach the servers from windows? (Just grasping at straws here)

**Comment by samiralavi** [comment]
In the latest updates of Windows 11, the `experimental` configs mentioned above are moved to the main `wsl2` section in the `.wslconfig` file (not experimental anymore). The following worked for me, using Ubuntu 22 and Cisco AnyConnect VPN:
```
[wsl2]
networkingMode=mirrored
dnsTunneling=true
```

**Comment by roadSurfer** [comment]
Looks like this can also be set via the UI.

![Image](https://github.com/user-attachments/assets/0aa0deaf-212b-41bb-a7bf-acb5d31a7956)

**Comment by vec715** [comment]
omg they I can't believe that MS STILL cant fix this hell issue ... im done with this crap

**Comment by marbaa** [comment]
Out of curiosity. Does anybody uses two VPN's and mirrored mode?

1st VPN to connect to corporate network (for example Cisco AnyConnect)
2nd VPN to connect to particular network segments (for example CheckPoint Securemote)

On my computer ssh doesn't work from WSL2 to networks behind 2nd VPN, however ssh works from Windows CMD to networks behind 2nd VPN. Same routes are visible in WSL2 and in Windows.

edit:
Just made a test in office, where no 1st VPN is active. SSH doesn't work from mirrored WSL2 to networks behind Checkpoint VPN's. So probably WSL2 is not compatible with Securemote Checkpoint VPN.

**Comment by martencassel** [comment]
Sayonara issue

**Comment by jpbuecken** [comment]
but why?

**Comment by gmcnitt** [comment]
[experimental]
networkingMode=mirrored
dnsTunneling=true

solved my issue 100%!  Cisco VPN works now!

**Comment by jpbuecken** [comment]
> [experimental] networkingMode=mirrored dnsTunneling=true
> 
> solved my issue 100%! Cisco VPN works now!

The solution still has issues when IPv6 is disabled. There are some issues here that are closed and resolved by enabling IPv6.
But some companies may have policies in place that do not allow to enable IPv6 on your system.

And in my case /etc/resolv.conf will become a dead symlink and DNS resolution does not work at all.

Maybe both are separate issue, but sub tasks to archive "Full VPN Support with WSL2" as the title of this issue suggests.

**Comment by shyney7** [comment]
> Looks like this can also be set via the UI.
> 
> ![Image](https://github.com/user-attachments/assets/0aa0deaf-212b-41bb-a7bf-acb5d31a7956)

This settings helped me thanks! Was having to deal with this issue since 2020 after 5 years microsoft seems to have managed to fix this...

**Comment by yttriumz** [comment]
> [experimental]
> networkingMode=mirrored
> dnsTunneling=true

Unfortunately these two settings do not make OpenVPN (v2.6.14) work together with Fedora 42 installed in WSL 2.

**Comment by muratkucuktepe** [comment]
My podman containers communicate with my host apps if I use "Mirrored" networking mode but WSL can not reach my company's domains.

My podman containers do not communicate with my host apps if I use "Nat" networking mode but WSL can reach my company's domains. 

I connect with VPN. Still could not find how I can connect to company's links via VPN and "Mirrored" mode.

**Comment by jpbuecken** [comment]
@muratkucuktepe : First step you can check
```
ip a
```
and
```
cat /etc/resolv.conf
```

with both modes. Does it look similar? (Number of network interfaces, same search domain in resolv.conf etc).
With mirrored mode, I only have a lo interface and a resolv.conf does not exist. In that case I assume it is the IPv6 issue I mentioned above.

**Comment by leo60228** [comment]
> The solution still has issues when IPv6 is disabled. There are some issues here that are closed and resolved by enabling IPv6.

@jpbuecken [Per Microsoft](https://learn.microsoft.com/en-us/troubleshoot/windows-server/networking/configure-ipv6-in-windows), disabling IPv6 is an unsupported configuration that may break OS functionality, and should only be done temporarily for troubleshooting purposes. Presumably, this implies that bugs that only occur when IPv6 is disabled are invalid.

**Comment by jpbuecken** [comment]
I have to admit I was not aware of that article. It may become handy.

I assumed IPv6 is disabled due to security concerns.

Unfortunately, Microsoft is inconsistent in its recommendations.
For example, they recommend disabling IPv6 as a mitigation for security issues like https://msrc.microsoft.com/update-guide/vulnerability/CVE-2024-38063
And I assume there are some more concerns behind it like https://github.com/microsoft/WSL/issues/11002#issuecomment-1883693635

The problem is the security argument is almighty unless it can be lifted. At least currently as it looks like most services for the daily workload work flawlessly. So WSL with VPN could be considered an edge case in comparison to a whole company.

Long story short, it would still be nice to support VPN / mirrored mode in environments with ipv6 disabled, even if it is not a recommended setting in general.

**Comment by paulmedynski** [comment]
I ran into this problem yesterday when using the Windows built-in VPN.  I got things to work by following these steps:

1. Shutdown WSL.
2. Enable the following WSL network settings:
  - Networking mode: NAT
  - Auto proxy enabled: On
  - DNS proxy enabled: On
3. Connect to VPN.
4. Start WSL session.

Those steps allow my WSL session (Debian 12) to connect to resources via the Windows VPN.  Connecting to the VPN _after_ starting a WSL session doesn't work.

**Comment by nithinjp1997** [comment]
> I've struggled with this for several months. It seems that a recent [update ](https://devblogs.microsoft.com/commandline/windows-subsystem-for-linux-september-2023-update/)to WSL offers experimental features which resolve the issue for me. I'm using WSL2 on Windows 11 with Cisco AnyConnect VPN. In my environment, I've added the following flags to my `C:\Users\_username_\.wslconfig` file:
> 
> ```
> [experimental]
> networkingMode=mirrored
> dnsTunneling=true
> ```

This fixed it for me. Just an update now these settings have been moved out of experimental so instead [experimental] use [wsl2] everything else remains same.

**Comment by paulmedynski** [comment]
FYI, I see this message currently when launching my WSL Debian distro for the first time after a shutdown:
```
wsl: Due to a current compatibility issue with Global Secure Access Client, DNS Tunneling is disabled.
```
I'm not sure if that will affect if/how Debian traffic would get routed via the VPN.  I'll report back the next time I need to do that.
I believe Global Secure Access Client is the built-in Windows VPN for AAD/Entra based authentication.

I'm up-to-date with my WSL setup:
```
WSL version: 2.6.1.0
Kernel version: 6.6.87.2-1
WSLg version: 1.0.66
MSRDC version: 1.2.6353
Direct3D version: 1.611.1-81528511
DXCore version: 10.0.26100.1-240331-1435.ge-release
Windows version: 10.0.26200.6584
```

**Comment by adriantorrie** [comment]
> > I've struggled with this for several months. It seems that a recent [update ](https://devblogs.microsoft.com/commandline/windows-subsystem-for-linux-september-2023-update/)to WSL offers experimental features which resolve the issue for me. I'm using WSL2 on Windows 11 with Cisco AnyConnect VPN. In my environment, I've added the following flags to my `C:\Users\_username_\.wslconfig` file:
> > ```
> > [experimental]
> > networkingMode=mirrored
> > dnsTunneling=true
> > ```
> 
> This fixed it for me. Just an update now these settings have been moved out of experimental so instead [experimental] use [wsl2] everything else remains same.

Confirmed on Cisco Secure Client.

[Microsoft docs reflect this guidance](https://learn.microsoft.com/en-us/windows/wsl/networking#mirrored-mode-networking) and the `.wslconfig` [settings docs confirm](https://learn.microsoft.com/en-us/windows/wsl/wsl-config#main-wsl-settings) it as well.

File: `c:\Users\<user_name>\.wslconfig`

```sh
# References
# - [wsl2] section
#   - https://learn.microsoft.com/en-us/windows/wsl/wsl-config#configuration-settings-for-wslconfig
# - [experimental] section
#   - https://learn.microsoft.com/en-us/windows/wsl/wsl-config#experimental-settings
[wsl2]
networkingMode=mirrored
```

**Comment by defrancr** [comment]
In my case using Cisco Secure Client, AnyConnect VPN 5.1.7.80, WSL version: 2.6.1.0 with Ubuntu 24.04, the .wslconfig that did the trick was:
```
[wsl2]
...
networkingMode=VirtioProxy
dnsTunneling=true
dnsProxy=false
```
(N.B.: Not sure if all the three parameters are mandatory to have it working)

**Comment by hellokatili** [comment]
@defrancr Not sure about the other settings but switching from `mirrored` to `VirtioProxy` worked for me.

Did not work:
```
[wsl2]
...
networkingMode=mirrored
dnsTunneling=true
dnsProxy=false
```

Works:
```
[wsl2]
...
networkingMode=VirtioProxy
dnsTunneling=true
dnsProxy=false
```

**Comment by jpbuecken** [comment]
@defrancr Worked for me as well. Great thanks: Switched from NAT to VirtioProxy.

For information: I have two VPNs , both configured in the same VPN client software. One worked (with networkingMode NAT), the other not, untill today. Both work with VirtioProxy.

I checked the difference in the vpn client config file and found nothing that explains why one worked and the other not.
So I assume it is not a client software issue per se, but different configurations on the VPN endpoint side are relevant as well.

**Comment by mloskot** [comment]
@hellokatili in https://github.com/microsoft/WSL/issues/10380#issuecomment-4175373364
> @defrancr Not sure about the other settings but switching from `mirrored` to `VirtioProxy` worked for me.

Does the VS Code work for you with this networking configuration? i.e. opening folder inside WSL with VS Code

This issue may be related 
- https://github.com/microsoft/WSL/issues/11184

**Comment by jpbuecken** [comment]
@mloskot It did not work in my Visual Studio Code. Pop up with socket error that 127.0.0.1:<port> cannot be reached.

It worked after adding `localhostForwarding=false` in my .wslconfig.

Little bit ironic since this option disables communication with ports on 127.0.0.1...

So this is everything in my .wslconfig ( WSL Version 2.6.3.0)
```
[wsl2]
networkingMode=VirtioProxy
localhostForwarding=false
```

It looks like dnsTunneling=true is default with most recent WSL (In WSL settings GUI app, it is active)

**Comment by mloskot** [comment]
@jpbuecken I'm happy to confirm the `localhostForwarding=false` was the missing piece indeed. This is great and thank you very much for sharing your solution!

**Comment by spanishkangaroo** [comment]
In my case, for **Ivanti Secure Access Client** (Pulse Secure / Juniper, sorry for the Spanish):

<img width="1672" height="940" alt="Image" src="https://github.com/user-attachments/assets/0eab92d9-3afb-4870-a5b3-690c45cca5a8" />

Also, ensure resolv.conf generation is active: inside the WSL machine,` /etc/resolv.conf` should contain:

```
# This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
# [network]
# generateResolvConf = false
nameserver <yourprivate-or-company-domain-server.ip>
search <yourprivate-or-company-domain.whatever> home
```

Just my two cents, as I arrived in 2026 to this thread and seems still relevant.
@gwd666

**Comment by AThomsen** [comment]
> [@jpbuecken](https://github.com/jpbuecken) I'm happy to confirm the `localhostForwarding=false` was the missing piece indeed. This is great and thank you very much for sharing your solution!

Unfortunately this breaks Podman Desktop :-/

**Comment by gwd666** [comment]
btw ... just a small comment on this topic ... 
example setup or 'what currently works for me' (with Pulse Secure)
It's having two "configurations" (one called for example `.wslconfig.home`, and one `.wslconfig.onsite`) in two files 
e.g. placed in my home/userprofile folder.
One is intended to be active when working from home - containing: `networkdingMode=mirrored` and `autoProxy=false` 
and one when working onsite which has only one networking setting defined i.e.: `networkingMode=nat` (which afaik is the "default" anyhow, so it could theoretically be omitted, but I put it in there nevertheless). 
Why? I learned/noticed that a working home-office config will break when onsite and vice-versa, so I played around until I found the sweet-spot regarding the `networkingMode/s`.
Then I use a simple powershell script (pwsh only b/c it's my preferred terminal on the Windows host), which will accept a <location> parameter (either 'home' or 'onsite'), to move/replace one of the two .wslconfig.<localtion>  files over to the `.wslconfig` 'place' (meaning copy/overwriting it). The final line in the script executes then `wsl --shutdown` such that WSL2 picks up the new `.wslconfig`, and that's it.  
On top of it - I am running a `ntlm` proxy (in WSL2 on localhost:port) which in my case is a pre-requisite (in both locations), so that I'm able to get through (to) the corp proxy (b/c that uses ntlm auth) - but that is a different story.   
It's a solution I've been able to live with for quite some time now, since it means I only need to run that script once, when switching my workplace from home to office or back. 
But lets see how long it will stay that way or some change/update (in its chain) breaks it 🚱 
The issues with loosing network connection (at home) and then (when reconnection happens) this 'bombing' the WSL network connectivity (only being recovarable with a reboot), still occurs but way less often; but tbh I can't identify if that is due to SW updates having improved network connection stability or if my ISP did something along those lines, or if it is a result of these new setup/network modes, or the move from WIN10 to WIN11 (WIN11 being a pre-requisite for all those additional network modes anyway), or a combo of all or some of them, etc. 🤷‍♂️
